### DISCUSSION TITLE
Liquidsoap hogging CPU

### DISCUSSION THREAD

**Discussion by Tycho-X** [AzuraCast/AzuraCast#6573] [Q&A] [answered]
Hello,

i did made my first installation of AzureCast on a Debian 11 VM. Installation went smooth but there is a major problem with the LiquidSoap processes. On the azurecast instannce there is only one active test stream and a total of 2 streams configured. Still there are 2 liquidsoap processes constantly hogging CPU between 40% and more often, 70-8%. Even when 100% / without any listeners. I installed the "rolling release" because this had been the default. 

Any tips on how to debug this? The liquidsoap log is empty and i cant find anything in the other logfiles on the VM. Also is there a quick way to install some basic tools like "less" and "vi" on the azurecast container?

Should i try to downgrade to stable? Can i downgrade from current "rolling release" to stable?

cheers.... 

![screeny](https://github.com/AzuraCast/AzuraCast/assets/9759464/8de064a1-ab56-4bd4-8019-ca1677a0506c)

**Comment by Moonbase59** [comment, marked as answer]
In `/var/azuracast/azuracast.env` on the host:

```
# Extra Ubuntu packages to install upon startup
# Standard: "Separate package names with a space. Packages will be installed during container startup."
INSTALL_PACKAGES_ON_STARTUP=nano gdb
```
Just packages that are in the normal repos work, and they might not have full functionality (locale, systemd, etc.). If this doesn’t work, you can always open a bash _inside the docker_, refresh the package list, and install temporarily (won’t be kept after the next `docker compose down`):

```
cd /var/azuracast
sudo ./docker.sh bash 
```
and inside the container:
```
sudo apt update
sudo apt install nano gdb
```

HTH.

Re downgrading: Depends. I wouldn’t—Rolling Release is currently a bumpy road, due to many changes and a frontend rework, but it’s rather stable and has so much new functionality.

Re CPU hogs: In _Edit Station Profile → AutoDJ → Advanced Configuration_, have you _Use ReplayGain Metadata_ enabled and many files _without_ ReplayGain tags? A _real CPU hog_, since Liquidsoap will try to calculate RG for files that don’t have it, and that comes in costly. Better pre-gain all your music, then use `radio = amplify(1.,override="replaygain_track_gain",radio)` in your Liquidsoap config (in the box before the crossfade code). This will _read and apply_ existing RG metadata, but not try to _calculate_ RG.

Another CPU hog can be StereoTool, should you use that. Should you need a mastering tool (you shouldn’t), _master_me_ is much more resource-friendly.

Also, watch Steal and I/O wait (Admin panel or `top` (`st`, `wa`)). Values not near zero here can indicate a cloud VM provider who is over-provisioning. Get them not to do that, move you to another server, or terminate the contract.

**Comment by Tycho-X** [comment]
Hello,

thank you for your answer. In the meantime i did make a new install on a new VM with the stable version. The stable versions CPU usage had been a lot better from the start, about 1/3rd of what the dev version used.

Now we also disabled replay gain and CPU usage is kinda ok. Thank you...

best regards
